src/Compiler/Microsoft.AspNetCore.Razor.Language/src/AllowedChildTagDescriptor.cs

```diff
@@ -1,75 +1,48 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#nullable disable
-
 using System;
-using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Linq;
 
 namespace Microsoft.AspNetCore.Razor.Language;
 
-public abstract class AllowedChildTagDescriptor : IEquatable<AllowedChildTagDescriptor>
+public sealed class AllowedChildTagDescriptor : TagHelperObject, IEquatable<AllowedChildTagDescriptor>
 {
-    private bool _containsDiagnostics;
-
-    public string Name { get; protected set; }
-
-    public string DisplayName { get; protected set; }
+    public string Name { get; }
+    public string DisplayName { get; }
 
-    public IReadOnlyList<RazorDiagnostic> Diagnostics
+    internal AllowedChildTagDescriptor(string name, string displayName, ImmutableArray<RazorDiagnostic> diagnostics)
     {
-        get => _containsDiagnostics
-            ? TagHelperDiagnostics.GetDiagnostics(this)
-            : Array.Empty<RazorDiagnostic>();
+        Name = name;
+        DisplayName = displayName;
 
-        protected set
+        if (!diagnostics.IsDefaultOrEmpty)
         {
-            if (value?.Count > 0)
-            {
-                TagHelperDiagnostics.AddDiagnostics(this, value);
-                _containsDiagnostics = true;
-            }
-            else if (_containsDiagnostics)
-            {
-                TagHelperDiagnostics.RemoveDiagnostics(this);
-                _containsDiagnostics = false;
-            }
+            SetFlag(ContainsDiagnosticsBit);
+            TagHelperDiagnostics.AddDiagnostics(this, diagnostics);
```

Comment:
I am so skeptical that maintaining a CWT for these is a worthwhile savings, but that's a refactoring for a future PR I think.